File: exercicios/projeto-final/skeletons/studio-ghibli/flutter/pratica/lib/api/ghibli_api.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/film.dart';

class GhibliApi {
  static const String baseUrl = 'https://ghibliapi.vercel.app';

  Future<List<FilmSummary>> fetchList() async {
    final res = await http.get(Uri.parse('$baseUrl/films'));
    if (res.statusCode != 200) {
      throw Exception('Falha ao carregar lista de filmes');
    }
    final body = jsonDecode(res.body) as List;
    return body.map((e) => FilmSummary.fromJson(e as Map<String, dynamic>)).toList();
  }

  Future<FilmDetail> fetchDetail(String id) async {
    final res = await http.get(Uri.parse('$baseUrl/films/$id'));
    if (res.statusCode != 200) {
      throw Exception('Falha ao carregar detalhe do filme $id');
    }
    return FilmDetail.fromJson(jsonDecode(res.body) as Map<String, dynamic>);
  }

  /// A própria API filtra por diretor — sem precisar de endpoint separado
  /// de "categoria" como o PokeAPI.
  Future<Set<String>> fetchTitlesByDirector(String director) async {
    final res = await http.get(Uri.parse('$baseUrl/films?director=$director'));
    if (res.statusCode != 200) {
      throw Exception('Falha ao carregar diretor $director');
    }
    final body = jsonDecode(res.body) as List;
    return body.map((e) => e['title'] as String).toSet();
  }
}

File: exercicios/projeto-final/skeletons/studio-ghibli/react-native/pratica/src/api/ghibliApi.ts
import { FilmDetail, FilmSummary } from '../types/film';

const BASE_URL = 'https://ghibliapi.vercel.app';

export async function fetchList(): Promise<FilmSummary[]> {
  const res = await fetch(`${BASE_URL}/films`);
  if (!res.ok) throw new Error('Falha ao carregar lista de filmes');
  const body = await res.json();
  return body.map((entry: { id: string; title: string }) => ({
    id: entry.id,
    title: entry.title,
  }));
}

export async function fetchDetail(id: string): Promise<FilmDetail> {
  const res = await fetch(`${BASE_URL}/films/${id}`);
  if (!res.ok) throw new Error(`Falha ao carregar detalhe do filme ${id}`);
  return res.json();
}

// A própria API filtra por diretor — sem precisar de endpoint separado de
// "categoria" como o PokeAPI.
export async function fetchTitlesByDirector(director: string): Promise<Set<string>> {
  const res = await fetch(`${BASE_URL}/films?director=${encodeURIComponent(director)}`);
  if (!res.ok) throw new Error(`Falha ao carregar diretor ${director}`);
  const body = await res.json();
  return new Set(body.map((entry: { title: string }) => entry.title));
}
